Try forwarding repeat records immediately

kaapstorm opened the pull request:
Context: https://dimagi-dev.atlassian.net/browse/HI-451 and PR #23579 

We won't need to iterate new repeat records if we forward them (successfully) immediately.

@dannyroberts you may know why we didn't do this before. I couldn't think of a reason.

@dannyroberts cc @czue @orangejenny @millerdev

kaapstorm committed:
Try forwarding repeat records immediately

dannyroberts commented on corehq/motech/repeaters/models.py:
```diff
@@ -163,16 +164,27 @@ def register(self, payload, next_check=None):
             payload_id=payload.get_id
         )
         repeat_record.save()
+        if next_check is now:
```
Nit: I was briefly confused by this, but it seems that `is` is used instead of `==` to emphasize that this is checking whether `next_check` was just set to `now` in the if statement above, right? It might be a little more straightforward to keep using `next_check=next_check or now` above and check for `if next_check is None` on this line instead.

kaapstorm replied:
Lol, in fact I considered leaving the code as it was, but chose `if next_check is now` because I imagined it was more readable. I'll go with what you're suggesting.

dannyroberts reviewed (approved):
Hey @kaapstorm I think this should work fine as you say. I'm not sure there was ever a specific reason that I didn't do it. Seems this part of the code maybe just hasn't changed since I first threw this feature together a long time ago

millerdev reviewed (commented):
I'm not sure I've completely wrapped my brain around this yet, so most of my comments here are just me trying to figure out how this is working and brainstorming about how to resolve the huge number of lock acquisitions we are seeing in datadog.

I think the new `forward_now(repeat_record)` call means there will be even more already-locked records returned by the `iterate_repeat_records(start)` call in the `check_repeaters()` task. Does that sound right?

--

Trying to understand what `iterate_repeat_records(start)` is doing: it looks like it's selecting all records with a `next_check` date before `start` where `start` is set to `datetime.now()` at the time when `check_repeaters()` is called. So to get a rough order of operations here (example times added for clarity):

```py
# in Repeater.register()
now = datetime.utcnow()  # 2:33pm
repeat_record.next_check = now
# in forward_now(repeat_record)
repeat_record.redis_lock.acquire(blocking=False)  # lock for 48 hours
process_repeat_record.delay(repeat_record)

# some time later in check_repeaters()
# assume process_repeat_record(repeat_record) has not updated the doc yet
# this is apparently a common scenario, given the number of lock acquire
# attempts we're seeing in datadog
start = datetime.utcnow()  # 2:34pm
for record in iterate_repeat_records(due_before=start):
    # this loop will include the repeat_record that was created above,
    # which has been queued to be processed by forward_now
    if not record.redis_lock.acquire(blocking=False):
        # it will be caught here and ignored since it has
        # been locked for the next 48 hours
        continue
    ...
```

--

Now that we're doing the `forward_now()` trick, would it work to eliminate the 48h lock and set `next_check=now + timedelta(hours=48)`? That way `iterate_repeat_records()` would not return it to be processed again until the lock would have expired.

Outline of how I think this new algorithm will work:

- create new repeat record
- set record to be processed next in 48 hours (in case the next step fails)
- forward now
  - typically this sets record.success to true, which means it will be removed from the `repeaters/repeat_records_by_next_check` view
- 48 hours later if the record has not been updated it will be returned by `iterate_repeat_records()` and processed again. I suppose this should also increment `next_check` by another 48 hours.

I think this is better because it defers further processing for 48 hours rather than slamming the lock repeatedly until it is finally processed or the lock expires, whichever comes first. We might even be able to reduce the 48 hour waiting period. Ideally we could make it dynamic, something like twice the length of the current backlog, although I haven't thought about how we would calculate that yet.

kaapstorm commented on corehq/motech/repeaters/models.py:
```diff
@@ -163,16 +164,27 @@ def register(self, payload, next_check=None):
+
+            if not record.redis_lock.acquire(blocking=False):
+                return
+            process_repeat_record.delay(repeat_record)
```
I specifically used a different variable name to [avoid it shadowing](http://www.oznetnerd.com/python-shadowing/) `repeat_record` in the outer scope (as you mentioned in Book Club, the reason Python 3 introduced `nonlocal`.) What do you think of using the name "repeat_record_" in the inner scope to make it clear that it's the same thing, but still avoid shadowing the outer variable name? Or do you feel shadowing is a non-issue?

millerdev replied:
My preference here would be to use the same name in both places because it references _the same object_ in both places.

Interesting article. Of the three examples given, I think _Avoid shadowing by using functions_ was my favorite (although I woudln't necessarily employ it in the code we're reviewing here), followed closely by _It’s fine to use shadowing_. I think there are more important best practices, such as writing _short functions_ and having them _do one thing_, that can be followed to make it easier to spot naming errors. The gotcha scenario they outlined where `name` was renamed to `person` didn't really resonate because that's very close to what you were trying to do by applying the _Avoid shadowing by ensuring all names are unique_ rule to `forward_now`, but you still got bitten. In this case you've followed both of the other best practices I mentioned above, and without a doubt that helped me spot the error.

BTW, it still hasn't been fixed. That's my fault for combining two separate issues in the same comment—you fixed the nit, but not the more important issue.

kaapstorm commented:
@millerdev I like your algorithm a lot. I found the way repeat records are locked interesting, because, like you suggested, ideally we want to back off incrementally. And we don't want multiples of 24 in case the remote service is only online during fixed hours, and we're trying to forward a record outside those hours. So maybe we want to try, say, 5 hours later, then 10, then 20. 

Should repeat records be locked at all? We could set the lock period to be the same as the `next_check` interval. `next_check` has always been there, but I think the locking is more recent. **Is there a reason the lock was added?**

kaapstorm committed:
PR feedback

kaapstorm committed:
Lock check_repeaters til all records are iterated

kaapstorm committed:
Clean imports

kaapstorm committed:
What? No rot13, Python 3?!

kaapstorm commented:
It seems we do need to lock repeat records, because we want to separate the purpose of check_next from making sure we don't queue the same payload multiple times.

So instead I've used the `check_repeater_lock` to iterate all repeat records, without being interrupted. I'm hoping we can get through all repeaters within half the repeat record lock timeout (i.e. 24 hours!).

czue commented:
Haven't reviewed the code in detail yet, but just to confirm I understand the high level change - it is to attempt to send repeaters in the signal instead of asynchronously, right?

If that's correct, I would be quite cautious about rolling this out. Think you would be moving a whole bunch of work from the celery machines to the web machines - work that is doing lots of I/O to unreliable remote servers which makes it even more likely to act up. Repeater backlogs are bad, but using up all our web process resources would be much worse I think.

I'm pretty sure we put all repeater logic in celery because we didn't want it to potentially affect form submissions or interfere with the web processes, no? cc @dannyroberts

kaapstorm commented:
I totally agree @czue. This still uses Celery (`process_repeat_record.delay(repeat_record)` on [models.py:172](https://github.com/dimagi/commcare-hq/pull/23580/files#diff-037a3a5cce75269c3fdd6318acdd6fe1R172)) to send asynchronously. I just drops a task in the `CELERY_REPEAT_RECORD_QUEUE` immediately instead of waiting for the `check_repeaters()` task to get to its repeater and then do it.

czue commented:
Right, thanks :+1: . Sorry should have looked at the code.

millerdev commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -30,34 +36,27 @@
-        lock_key = _get_repeat_record_lock_key(record)
-
-        if now > cutoff:
+        if not _soft_assert(
```
nit: the negation makes it hard to understand the condition (I didn't even know `SoftAssert.__call__` returned the assertion condition). What do you think of restructuring this as:
```py
if datetime.utcnow() > timeout_dt:
    _soft_assert(False, "...")
    break
```

kaapstorm replied:
OK. Fair enough. I kinda love that soft_assert [can be used in conditionals](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/tests/test_soft_assert.py#L35) but grepping the codebase makes me think I'm alone in that.

millerdev commented:
> ... ideally we want to back off incrementally. And we don't want multiples of 24 in case the remote service is only online during fixed hours, and we're trying to forward a record outside those hours. So maybe we want to try, say, 5 hours later, then 10, then 20.

Yes, completely agree with all of that.

> It seems we do need to lock repeat records, because we want to separate the purpose of `check_next` from making sure we don't queue the same payload multiple times.

As I'm sure you know, there are two different locks in play here, one is `repeat_record.redis_lock` (used to be `lock`) and the other is `check_repeater_lock`. The one I am proposing eliminating with the new algorithm is `repeat_record.redis_lock`. It is my understanding that that (48h) lock is being used as a protection against sending the same payload twice, although since it is acquired and never released it either (a) is not necessary at all or (b) the timeout is assumed to be long enough that _an error is assumed to have occurred preventing the payload from being sent_ in any case where the lock times out and the repeat record is still in the queue. This is an unconventional locking pattern, and I think there are better ways to achieve the same result without it.

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -30,34 +36,27 @@
+    timeout_dt = start + timedelta(seconds=repeater_lock_timeout)
 
-    # Timeout for slightly less than periodic check
+    # Long timeout to allow all waiting repeat records to be iterated
```
Agreed. In fact this structure makes more sense regardless. In this case, does it make sense to have a timeout at all? Or is waiting a whole day to iterate all repeat records just ridiculous, and we should drop it to, say, 6 hours? (This guesswork drives me nuts. What about 24 hours, log how long it takes, and drop it to that + 50%?)

millerdev replied:
> In this case, does it make sense to have a timeout at all?

I think timeouts are always used on redis locks in case of a failure that causes the lock to not be released (interpreter crash where `finally` clause is not executed, for example). Since the lock state is external to the process we want to make sure it is definitely released at some point.

However, I would argue that we should never rely on timeouts as part of the program happy path. Instead, the timeout should never be hit except for very rare and unexpected error scenarios. Ideally such errors should report themselves loudly so we know when they are occurring. If the frequency is high it is a sign that we need to re-think the design or at the very least the timeout.

> Or is waiting a whole day to iterate all repeat records just ridiculous, and we should drop it to, say, 6 hours? (This guesswork drives me nuts. What about 24 hours, log how long it takes, and drop it to that + 50%?)

Agree about the guesswork driving nuts. Might be useful to go to voice and try to work out a design that balances timeout length with implementation complexity. What do you think?

kaapstorm committed:
soft_assert(False, ...)

kaapstorm committed:
Don't ignore LockError

kaapstorm committed:
Use can_enqueue() instead of redis_lock

When this is deployed, records that are enqueued and locked but
self.queued is False may be queued again. I think resending them
outweighs the hassle of implementing this change over two PRs.

kaapstorm committed:
Finally fixed variable name, ignoring shadowing

kaapstorm committed:
Drop lock timeout. Quit after 6 hours.

kaapstorm committed:
Use a six-hour timeout on check_repeater_lock

kaapstorm committed:
Implement original 48-hour lock using next_check

(instead of using a new "queued" property)

kaapstorm committed:
nit

millerdev commented on corehq/motech/repeaters/models.py:
```diff
@@ -773,6 +781,14 @@ def cancel(self):
+            # self is already in the queue, or it is not ready to be enqueued
+            return False
+        self.next_check += timedelta(hours=48)
+        self.save()
```
Should catch `ResourceConfict` (and `return False`) here since two updates to the same record revision will cause that and we don't want it to bubble up. This is the crux of switching to optimistic locking and eliminating the old Redis locks.

Double-check that `ResourceConfict` is the exception raised when an old revision of a document is saved. I'm not positive above that.

millerdev commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -30,40 +35,27 @@
 
-    # Timeout for slightly less than periodic check
+    # Long timeout to allow all waiting repeat records to be iterated
     check_repeater_lock = get_redis_lock(
```
I think this lock is no longer necessary. If only one process is running, then it's definitely not necessary. If there are two or more, and multiple processes call `record.can_enqueue()` for the same record at the same moment, only one will return true, and all others will fail with `ResourceConflict` and return false.

kaapstorm replied:
That sounds right, but wouldn't dropping the lock result in churn that using it avoids?

millerdev replied:
Context from offline discussion: we decided that it might be better to measure first and then drop the lock later after verifying it's no longer needed. Do you want to add datadog monitoring of things like the _number of records_, _length of time spent in the loop_, and _the number of times the lock cannot be acquired_ in this PR or a follow-up?

kaapstorm committed:
Catch ResourceConflict

kaapstorm committed:
Fix tests

millerdev commented on corehq/motech/repeaters/models.py:
```diff
@@ -773,6 +781,17 @@ def cancel(self):
         self.next_check = None
         self.cancelled = True
 
+    def can_enqueue(self):
```
This `can_enqueue` method makes me a little uncomfortable because it changes state even though it looks like a benign state check. If you call it twice quickly the second result will always be false.

What do you think of renaming it to `forward_now` and moving `process_repeat_record.delay(repeat_record)` to where it returns true?

```py
def forward_now(self):
    from corehq.motech.repeaters.tasks import process_repeat_record
    if self.next_check > datetime.utcnow():
        # self is already in the queue, or it is not ready to be enqueued
        return
    # Set the next check to happen an arbitrarily long time from now so
    # if something goes horribly wrong with the delayed task it will not
    # be lost forever. This check is expected to occur rarely, if ever,
    # because `process_repeat_record` will either succeed or will reset
    # the next check to something sooner.
    self.next_check = datetime.utcnow() + timedelta(hours=48)
    try:
        self.save()
    except ResourceConflict:
        # Another process beat us to the punch. This takes advantage of
        # the fact Couch uses optimistic locking to prevent one process
        # from overwriting the work of another with stale data.
        return
    process_repeat_record.delay(self)
```

As you can see I added some verbose comments so future readers can get a better handle on what's happening here. I also made another somewhat critical change to prevent a very old `next_check` value from being incremented to a time that is not two days from now:
```diff
- self.next_check += timedelta(hours=48)
+ self.next_check = datetime.utcnow() + timedelta(hours=48)`
```
The places that call `record.can_enqueue()` will need to be changed to call `record.forward_now()`

millerdev replied:
Maybe it should be named `attempt_forward_now`

kaapstorm replied:
Thank you. That not only addresses the issue of the state change, but also resolves something I was uncomfortable about; separating that state change from the purpose of the state change. It felt wrong to me that we were changing `next_check` to indicate that the record was queued separately from actually putting it in the repeat record queue. I'm glad this resolves both. I'll make the change.

kaapstorm committed:
process_repeat_record in can_enqueue, and rename

kaapstorm committed:
Fix tests

millerdev commented on corehq/motech/repeaters/models.py:
```diff
@@ -773,6 +775,27 @@ def cancel(self):
+    def attempt_forward_now(self):
+        from corehq.motech.repeaters.tasks import process_repeat_record
+
+        if self.next_check > datetime.utcnow():
```
Should this check these other things first, in case the record is in an unexpected state?
```py
if not (self.succeeded or self.cancelled or self.next_check is None):
    return
```

I think all of that is redundant with the state of newly created repeat records and the `repeaters/repeat_records_by_next_check` view definition, but those are sufficiently far away from this code that it may make sense to double-check here. What do you think?

kaapstorm replied:
I think it makes sense, even if only to help someone new to the code understand how this works. (Are you sure about that `not`?)

millerdev replied:
> Are you sure about that `not`?

nope, accidentally reversed the logic along the way somehow.

kaapstorm committed:
Add already_processed() check

kaapstorm committed:
Fix tests

kaapstorm committed:
Put repeat record redis lock back temporarily

kaapstorm commented:
Wanna check this one last time @millerdev ?

millerdev commented on corehq/motech/repeaters/models.py:
```diff
@@ -800,7 +817,8 @@ def already_processed():
             # from overwriting the work of another with stale data.
             return
-        process_repeat_record.delay(self)
+        if self.redis_lock.acquire():
```
Should this lock be acquired before `self.save()` is called? All of the existing locks are based on revisions of the record that have not been updated with the new `next_check` value, so `self.save()` will unlock the record because of

https://github.com/dimagi/commcare-hq/blob/776cdc50385a9669cae1c3f8673dbcb2abbc53e9/corehq/motech/repeaters/models.py#L785-L786

I had initially thought we may need to acquire it before _and after_ the save call, but I don't think it's necessary since the `next_check` date will be set in the future by `self.save()`.

millerdev replied:
A possibly better option would be to move `if self.redis_lock.acquire():` back into `check_repeaters` before

https://github.com/dimagi/commcare-hq/blob/776cdc50385a9669cae1c3f8673dbcb2abbc53e9/corehq/motech/repeaters/tasks.py#L54

I think that's the only place the lock is needed.

kaapstorm replied:
I think your initial thought is correct; It will need to block before `save()` otherwise it will never block, and it will need to be acquired after `save()` otherwise it won't block the next attempt.

I think it should be inside `attempt_forward_now()` because it also needs to be acquired when the repeat record is registered.

Urgh this is ugly. I'm glad it's temporary.

kaapstorm replied:
Context from offline discussion: @millerdev is correct; we only need to check the locks of repeat records whose `next_check` has not been set in the future, which registering and saving will do.

millerdev replied:
https://github.com/dimagi/commcare-hq/pull/23580/commits/09d4c092a6c73cdb8034cb2b2002ae0e2941a50b

kaapstorm replied:
Nice! :tada:

kaapstorm replied:
Thank you!

millerdev committed:
Temporarily reinstate the old lock

It's critical that the lock is acquired before next_check is reset and saved because the record will be unlocked on save.

millerdev committed:
Improve comment wording

kaapstorm commented:
... @millerdev made a good point offline: :thinking: It might be prudent to deploy this at the start of the week instead of Friday. Reverting, just in case.
